sys/kern/signal.c

```diff
@@ -173,17 +176,137 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
   return EINTR;
 }
 
+static ksiginfo_t *ksiginfo_copy(const ksiginfo_t *src) {
+  ksiginfo_t *kp = kmalloc(M_SIGNAL, sizeof(ksiginfo_t), M_NOWAIT);
+  assert(kp != NULL);
+
+  memcpy(kp, src, sizeof(ksiginfo_t));
+  kp->ksi_flags &= ~KSI_QUEUED;
+  kp->ksi_flags |= KSI_FROMPOOL;
+  return kp;
+}
+
+static void ksiginfo_free(ksiginfo_t *ksi) {
+  assert((ksi->ksi_flags & KSI_FROMPOOL) != 0);
+  assert((ksi->ksi_flags & KSI_QUEUED) == 0);
+
+  kfree(M_SIGNAL, ksi);
+}
+
+void sigpend_init(sigpend_t *sp) {
+  assert(sp != NULL);
+  TAILQ_INIT(&sp->sp_info);
+  __sigemptyset(&sp->sp_set);
+}
+
+void sigpend_destroy(sigpend_t *sp) {
+  assert(sp != NULL);
+  __sigemptyset(&sp->sp_set);
+
+  ksiginfo_t *ksi;
+  TAILQ_FOREACH (ksi, &sp->sp_info, ksi_list) {
+    TAILQ_REMOVE(&sp->sp_info, ksi, ksi_list);
+    assert(ksi->ksi_flags & KSI_FROMPOOL);
+    assert(ksi->ksi_flags & KSI_QUEUED);
+    ksiginfo_free(ksi);
+  }
+}
+
+static void sigpend_get(sigpend_t *sp, signo_t sig, ksiginfo_t *out) {
+  __sigdelset(&sp->sp_set, sig);
+
+  /* Find siginfo and copy it out. */
+  ksiginfo_t *ksi;
+  TAILQ_FOREACH (ksi, &sp->sp_info, ksi_list) {
+    if (ksi->ksi_signo == sig)
+      break;
+  }
+
+  if (!ksi) {
+    if (out) {
+      bzero(out, sizeof(ksiginfo_t));
+      out->ksi_signo = sig;
+      out->ksi_code = SI_NOINFO;
+    }
+    return;
+  }
+
+  TAILQ_REMOVE(&sp->sp_info, ksi, ksi_list);
+  assert(ksi->ksi_flags & KSI_FROMPOOL);
+  assert(ksi->ksi_flags & KSI_QUEUED);
+  ksi->ksi_flags &= ~KSI_QUEUED;
+  if (out) {
+    memcpy(out, ksi, sizeof(ksiginfo_t));
+    out->ksi_flags &= ~KSI_FROMPOOL;
+  }
+  ksiginfo_free(ksi);
+}
+
+static void sigpend_put(sigpend_t *sp, ksiginfo_t *ksi) {
+  assert(sp != NULL);
+  assert(ksi != NULL);
+  assert((ksi->ksi_flags & KSI_QUEUED) == 0);
+
+  signo_t sig = ksi->ksi_signo;
+  __sigaddset(&sp->sp_set, sig);
+
+  assert(ksi->ksi_flags & KSI_FROMPOOL);
+
+  /* If there is no siginfo, we are done. */
+  if (ksi->ksi_flags & KSI_RAW) {
+    ksiginfo_free(ksi);
+    return;
+  }
+
+  ksiginfo_t *kp;
+  TAILQ_FOREACH (kp, &sp->sp_info, ksi_list) {
+    if (kp->ksi_signo == sig)
+      break;
+  }
+
+  if (kp) {
+    kp->ksi_info = ksi->ksi_info;
+    kp->ksi_flags = ksi->ksi_flags | KSI_QUEUED;
```

Comment:
You're leaking `ksi` here, you should `ksiginfo_free` it.

Reply:
Thanks, I missed it.